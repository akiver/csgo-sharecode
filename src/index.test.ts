import { describe, it, expect } from 'vitest';
import {
  InvalidShareCode,
  MatchInformation,
  decodeMatchShareCode,
  encodeMatch,
  InvalidCrosshairShareCode,
  Crosshair,
  decodeCrosshairShareCode,
  encodeCrosshair,
  crosshairToConVars,
} from '.';

const invalidShareCodes = [
  'CSGO-12345-12345-12345-12345-1234',
  'whateverCSGO-12345-12345-12345-12345-12345',
  'CSGO-12345-12345-12345-12345-12345whatever',
];

describe('Match share code', () => {
  const matchSamples: Array<{ shareCode: string; matchInformation: MatchInformation }> = [
    {
      shareCode: 'CSGO-L9spZ-ihuov-cyhtE-kxbqa-FkBAA',
      matchInformation: {
        matchId: BigInt('3400360672356205056'),
        reservationId: BigInt('3400367402569957763'),
        tvPort: 9725,
      },
    },
    {
      shareCode: 'CSGO-GADqf-jjyJ8-cSP2r-smZRo-TO2xK',
      matchInformation: {
        matchId: BigInt('3230642215713767580'),
        reservationId: BigInt('3230647599455273103'),
        tvPort: 55788,
      },
    },
    {
      shareCode: 'CSGO-bPQEz-PrYTq-u5w8E-ZbUy7-ZeQ3A',
      matchInformation: {
        matchId: BigInt('3325408798641750542'),
        reservationId: BigInt('3325410334092558852'),
        tvPort: 240,
      },
    },
    {
      shareCode: 'CSGO-wBrm6-7fkM6-AzBC5-u6GmR-iHLHA',
      matchInformation: {
        matchId: BigInt('3302232779302895618'),
        reservationId: BigInt('3302241568953467250'),
        tvPort: 3085,
      },
    },
    {
      shareCode: 'CSGO-TKDTJ-YrAXs-sDNfL-HOuKO-i84VH',
      matchInformation: {
        matchId: BigInt('3402250361329680757'),
        reservationId: BigInt('3402250801563828781'),
        tvPort: 61630,
      },
    },
    {
      shareCode: 'CSGO-p4X9o-3Mfut-tpe5y-J8K6f-mj5ZJ',
      matchInformation: {
        matchId: BigInt('3402249502336221574'),
        reservationId: BigInt('3402252092201501292'),
        tvPort: 14119,
      },
    },
  ];

  it('should decode', () => {
    matchSamples.forEach(({ shareCode, matchInformation }) => {
      expect(decodeMatchShareCode(shareCode)).toEqual(matchInformation);
    });
  });

  it('should encode', () => {
    matchSamples.forEach(({ shareCode, matchInformation }) => {
      expect(encodeMatch(matchInformation)).toEqual(shareCode);
    });
  });

  it('should throw an error if the share code is invalid', () => {
    invalidShareCodes.forEach((shareCode) => {
      expect(() => {
        decodeMatchShareCode(shareCode);
      }).toThrow(new InvalidShareCode());
    });
  });
});

describe('Crosshair share code', () => {
  const crosshairSamples: Array<{ shareCode: string; crosshair: Crosshair }> = [
    {
      shareCode: 'CSGO-Cn37R-YE7vo-pLCAL-aURmZ-z6zkG',
      crosshair: {
        gap: -1.3,
        outline: 2,
        red: 175,
        green: 81,
        blue: 213,
        alpha: 137,
        splitDistance: 6,
        fixedCrosshairGap: 3,
        color: 5,
        innerSplitAlpha: 0.6,
        outlineEnabled: true,
        outerSplitAlpha: 0.4,
        splitSizeRatio: 0.5,
        thickness: 1.2,
        centerDotEnabled: true,
        alphaEnabled: true,
        tStyleEnabled: true,
        style: 2,
        length: 4.6,
        deployedWeaponGapEnabled: true,
        followRecoil: false,
      },
    },
    {
      shareCode: 'CSGO-LibdP-VCVEd-ESayK-rSivi-2UBtG',
      crosshair: {
        gap: 1,
        outline: 1,
        red: 50,
        green: 250,
        blue: 50,
        alpha: 200,
        splitDistance: 7,
        fixedCrosshairGap: 3,
        color: 1,
        innerSplitAlpha: 1,
        outlineEnabled: false,
        outerSplitAlpha: 0.5,
        splitSizeRatio: 0.3,
        thickness: 0.5,
        centerDotEnabled: false,
        alphaEnabled: false,
        tStyleEnabled: false,
        style: 3,
        length: 5,
        deployedWeaponGapEnabled: false,
        followRecoil: false,
      },
    },
    {
      shareCode: 'CSGO-9JzcN-4dZtA-DdXis-8qz5T-rCnkP',
      crosshair: {
        gap: 1,
        outline: 1,
        red: 50,
        green: 250,
        blue: 50,
        alpha: 200,
        splitDistance: 7,
        fixedCrosshairGap: 4.3,
        color: 5,
        innerSplitAlpha: 1,
        outlineEnabled: false,
        outerSplitAlpha: 0.5,
        splitSizeRatio: 0.3,
        thickness: 0.5,
        centerDotEnabled: true,
        alphaEnabled: true,
        tStyleEnabled: false,
        style: 1,
        length: 5,
        deployedWeaponGapEnabled: true,
        followRecoil: false,
      },
    },
    {
      shareCode: 'CSGO-fCUBz-CBHss-a74RP-SEdO8-mvZpG',
      crosshair: {
        gap: 1,
        outline: 1,
        red: 50,
        green: 250,
        blue: 50,
        alpha: 200,
        splitDistance: 7,
        fixedCrosshairGap: 4.3,
        color: 5,
        innerSplitAlpha: 1,
        outlineEnabled: false,
        outerSplitAlpha: 0.5,
        splitSizeRatio: 0.3,
        thickness: 0.5,
        centerDotEnabled: true,
        alphaEnabled: true,
        tStyleEnabled: false,
        style: 2,
        length: 5,
        deployedWeaponGapEnabled: true,
        followRecoil: true,
      },
    },
    {
      shareCode: 'CSGO-WsnnD-eHaMw-QNDf9-oxuDh-ydOUD',
      crosshair: {
        gap: -2.2,
        outline: 1,
        red: 50,
        green: 250,
        blue: 50,
        alpha: 200,
        splitDistance: 3,
        fixedCrosshairGap: 3,
        color: 1,
        innerSplitAlpha: 0,
        outlineEnabled: true,
        outerSplitAlpha: 1,
        splitSizeRatio: 1,
        thickness: 0.6,
        centerDotEnabled: false,
        alphaEnabled: true,
        tStyleEnabled: false,
        style: 2,
        length: 10,
        deployedWeaponGapEnabled: true,
        followRecoil: true,
      },
    },
    {
      shareCode: 'CSGO-ZrEjo-yASEP-OAdce-Sf44w-rhK5O',
      crosshair: {
        gap: -1.2,
        outline: 1,
        red: 232,
        green: 88,
        blue: 227,
        alpha: 136,
        splitDistance: 7,
        fixedCrosshairGap: 3,
        color: 5,
        innerSplitAlpha: 0.5,
        outlineEnabled: false,
        outerSplitAlpha: 0.6,
        splitSizeRatio: 0.5,
        thickness: 1.5,
        centerDotEnabled: true,
        alphaEnabled: false,
        tStyleEnabled: true,
        style: 2,
        length: 7.2,
        deployedWeaponGapEnabled: false,
        followRecoil: true,
      },
    },
  ];

  it('should decode', () => {
    crosshairSamples.forEach(({ shareCode, crosshair }) => {
      expect(decodeCrosshairShareCode(shareCode)).toEqual(crosshair);
    });
  });

  it('should encode', () => {
    crosshairSamples.forEach(({ shareCode, crosshair }) => {
      expect(encodeCrosshair(crosshair)).toEqual(shareCode);
    });
  });

  it('should generate ConVars from crosshair', () => {
    crosshairSamples.forEach(({ crosshair }) => {
      expect(crosshairToConVars(crosshair)).toMatchSnapshot();
    });
  });

  it('should throw an error if the share code is invalid', () => {
    invalidShareCodes.forEach((shareCode) => {
      expect(() => {
        decodeCrosshairShareCode(shareCode);
      }).toThrow(new InvalidShareCode());
    });
  });

  it('should throw an error if the crosshair share code is invalid', () => {
    const invalidCrosshairCodes = [
      'CSGO-L9spZ-ihuov-cyhtE-kxbqa-FkBAA',
      'CSGO-12345-12345-12345-12345-12345',
      'CSGO-11111-22222-33333-44444-55555',
    ];

    invalidCrosshairCodes.forEach((shareCode) => {
      expect(() => {
        decodeCrosshairShareCode(shareCode);
      }).toThrow(new InvalidCrosshairShareCode());
    });
  });
});
