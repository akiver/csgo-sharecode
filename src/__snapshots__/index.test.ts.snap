// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Crosshair share code > should generate ConVars from crosshair 1`] = `
"
cl_crosshair_drawoutline \\"1\\"
cl_crosshair_dynamic_maxdist_splitratio \\"0.5\\"
cl_crosshair_dynamic_splitalpha_innermod \\"0.6\\"
cl_crosshair_dynamic_splitalpha_outermod \\"0.4\\"
cl_crosshair_dynamic_splitdist \\"6\\"
cl_crosshair_outlinethickness \\"2\\"
cl_crosshair_t \\"1\\"
cl_crosshairalpha \\"137\\"
cl_crosshaircolor \\"5\\"
cl_crosshaircolor_b \\"213\\"
cl_crosshaircolor_g \\"81\\"
cl_crosshaircolor_r \\"175\\"
cl_crosshairdot \\"1\\"
cl_crosshairgap \\"-1.3\\"
cl_crosshairgap_useweaponvalue \\"1\\"
cl_crosshairsize \\"4.6\\"
cl_crosshairstyle \\"2\\"
cl_crosshairthickness \\"1.2\\"
cl_crosshairusealpha \\"1\\"
cl_fixedcrosshairgap \\"3\\"
cl_crosshair_recoil \\"0\\"
"
`;

exports[`Crosshair share code > should generate ConVars from crosshair 2`] = `
"
cl_crosshair_drawoutline \\"0\\"
cl_crosshair_dynamic_maxdist_splitratio \\"0.3\\"
cl_crosshair_dynamic_splitalpha_innermod \\"1\\"
cl_crosshair_dynamic_splitalpha_outermod \\"0.5\\"
cl_crosshair_dynamic_splitdist \\"7\\"
cl_crosshair_outlinethickness \\"1\\"
cl_crosshair_t \\"0\\"
cl_crosshairalpha \\"200\\"
cl_crosshaircolor \\"1\\"
cl_crosshaircolor_b \\"50\\"
cl_crosshaircolor_g \\"250\\"
cl_crosshaircolor_r \\"50\\"
cl_crosshairdot \\"0\\"
cl_crosshairgap \\"1\\"
cl_crosshairgap_useweaponvalue \\"0\\"
cl_crosshairsize \\"5\\"
cl_crosshairstyle \\"3\\"
cl_crosshairthickness \\"0.5\\"
cl_crosshairusealpha \\"0\\"
cl_fixedcrosshairgap \\"3\\"
cl_crosshair_recoil \\"0\\"
"
`;

exports[`Crosshair share code > should generate ConVars from crosshair 3`] = `
"
cl_crosshair_drawoutline \\"0\\"
cl_crosshair_dynamic_maxdist_splitratio \\"0.3\\"
cl_crosshair_dynamic_splitalpha_innermod \\"1\\"
cl_crosshair_dynamic_splitalpha_outermod \\"0.5\\"
cl_crosshair_dynamic_splitdist \\"7\\"
cl_crosshair_outlinethickness \\"1\\"
cl_crosshair_t \\"0\\"
cl_crosshairalpha \\"200\\"
cl_crosshaircolor \\"5\\"
cl_crosshaircolor_b \\"50\\"
cl_crosshaircolor_g \\"250\\"
cl_crosshaircolor_r \\"50\\"
cl_crosshairdot \\"1\\"
cl_crosshairgap \\"1\\"
cl_crosshairgap_useweaponvalue \\"1\\"
cl_crosshairsize \\"5\\"
cl_crosshairstyle \\"1\\"
cl_crosshairthickness \\"0.5\\"
cl_crosshairusealpha \\"1\\"
cl_fixedcrosshairgap \\"4.3\\"
cl_crosshair_recoil \\"0\\"
"
`;

exports[`Crosshair share code > should generate ConVars from crosshair 4`] = `
"
cl_crosshair_drawoutline \\"0\\"
cl_crosshair_dynamic_maxdist_splitratio \\"0.3\\"
cl_crosshair_dynamic_splitalpha_innermod \\"1\\"
cl_crosshair_dynamic_splitalpha_outermod \\"0.5\\"
cl_crosshair_dynamic_splitdist \\"7\\"
cl_crosshair_outlinethickness \\"1\\"
cl_crosshair_t \\"0\\"
cl_crosshairalpha \\"200\\"
cl_crosshaircolor \\"5\\"
cl_crosshaircolor_b \\"50\\"
cl_crosshaircolor_g \\"250\\"
cl_crosshaircolor_r \\"50\\"
cl_crosshairdot \\"1\\"
cl_crosshairgap \\"1\\"
cl_crosshairgap_useweaponvalue \\"1\\"
cl_crosshairsize \\"5\\"
cl_crosshairstyle \\"2\\"
cl_crosshairthickness \\"0.5\\"
cl_crosshairusealpha \\"1\\"
cl_fixedcrosshairgap \\"4.3\\"
cl_crosshair_recoil \\"1\\"
"
`;

exports[`Crosshair share code > should generate ConVars from crosshair 5`] = `
"
cl_crosshair_drawoutline \\"1\\"
cl_crosshair_dynamic_maxdist_splitratio \\"1\\"
cl_crosshair_dynamic_splitalpha_innermod \\"0\\"
cl_crosshair_dynamic_splitalpha_outermod \\"1\\"
cl_crosshair_dynamic_splitdist \\"3\\"
cl_crosshair_outlinethickness \\"1\\"
cl_crosshair_t \\"0\\"
cl_crosshairalpha \\"200\\"
cl_crosshaircolor \\"1\\"
cl_crosshaircolor_b \\"50\\"
cl_crosshaircolor_g \\"250\\"
cl_crosshaircolor_r \\"50\\"
cl_crosshairdot \\"0\\"
cl_crosshairgap \\"-2.2\\"
cl_crosshairgap_useweaponvalue \\"1\\"
cl_crosshairsize \\"10\\"
cl_crosshairstyle \\"2\\"
cl_crosshairthickness \\"0.6\\"
cl_crosshairusealpha \\"1\\"
cl_fixedcrosshairgap \\"3\\"
cl_crosshair_recoil \\"1\\"
"
`;

exports[`Crosshair share code > should generate ConVars from crosshair 6`] = `
"
cl_crosshair_drawoutline \\"0\\"
cl_crosshair_dynamic_maxdist_splitratio \\"0.5\\"
cl_crosshair_dynamic_splitalpha_innermod \\"0.5\\"
cl_crosshair_dynamic_splitalpha_outermod \\"0.6\\"
cl_crosshair_dynamic_splitdist \\"7\\"
cl_crosshair_outlinethickness \\"1\\"
cl_crosshair_t \\"1\\"
cl_crosshairalpha \\"136\\"
cl_crosshaircolor \\"5\\"
cl_crosshaircolor_b \\"227\\"
cl_crosshaircolor_g \\"88\\"
cl_crosshaircolor_r \\"232\\"
cl_crosshairdot \\"1\\"
cl_crosshairgap \\"-1.2\\"
cl_crosshairgap_useweaponvalue \\"0\\"
cl_crosshairsize \\"7.2\\"
cl_crosshairstyle \\"2\\"
cl_crosshairthickness \\"1.5\\"
cl_crosshairusealpha \\"0\\"
cl_fixedcrosshairgap \\"3\\"
cl_crosshair_recoil \\"1\\"
"
`;
